kolena/workflow/io.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2021-2023 Kolena Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+from typing import Any
+from typing import Union
+
+import pandas as pd
+
+from kolena.workflow import DataObject
+from kolena.workflow._datatypes import _DATA_TYPE_MAP
+from kolena.workflow._datatypes import DATA_TYPE_FIELD
+
+
+def _serialize_dataobject(x: Any) -> Any:
+    if isinstance(x, list):
+        return [item._to_dict() if isinstance(item, DataObject) else item for item in x]
+
+    return x._to_dict() if isinstance(x, DataObject) else x
+
+
+def _deserialize_dataobject(x: Any) -> Any:
+    if isinstance(x, list):
+        return [_deserialize_dataobject(item) for item in x]
+
+    if isinstance(x, dict) and DATA_TYPE_FIELD in x:
+        data = {**x}
+        data_type = data.pop(DATA_TYPE_FIELD)
+        typed_dataobject = _DATA_TYPE_MAP.get(data_type, None)
+        if typed_dataobject:
+            return typed_dataobject._from_dict(data)
+
+    return x
+
+
+def _serialize_json(x: Any) -> Any:
+    if isinstance(x, list) or isinstance(x, dict):
+        return json.dumps(x)
+
+    return x
+
+
+def _deserialize_json(x: Any) -> Any:
+    if isinstance(x, str):
+        try:
+            return json.loads(x)
+        except Exception:
+            ...
+
+    return x
+
+
+def dataframe_to_csv(df: pd.DataFrame, *args, **kwargs) -> Union[str, None]:
```

Comment:
Does this mean that `df.to_csv` doesn't work? In what situations would it fail?

Reply:
`d.to_csv` would work in the sense that there would be no error writing source dataframe out to csv. However, the default serialization would be used, which I think is roughly `str`. For kolena dataobject in the dataframe, it would not be able to deserialize back, it would just be a string looking like 
```
"LabeledBoundingBox(top_left=(0.0, 0.0), bottom_right=(10.0, 10.0), width=10.0, height=10.0, area=100.0, aspect_ratio=1.0, label='foo-0')"
```

Reply:
[for dict in the df](https://github.com/kolenaIO/shipyard/pull/1570#discussion_r1330320818), it will not generate a valid json serialization when using `to_csv`, i think that's why we need to `df_post[columns] = df_post[columns].applymap(_serialize_json)`